### DISCUSSION TITLE
Needed help...local coordinate system for trackNode and mouse coordinates...

### DISCUSSION THREAD

**Discussion by keince** [pelicanmapping/osgearth#1954] [Support Q&A] [answered]
I'm in need of some help with a problem I'm trying to solve...

I have a trackNode that has just one drawable (no icon, no schema) which draws a single line segment. I set it this way so I can pick the line in the drawable of the trackNode. The first vertex of the line remains fixed at (0,0) in the local coordinate system of the trackNode and the second vertex is offset from the origin at (0,20). What I would like to do is pick the line (which I can do) and change the endpoint of the line by dragging it around using the mouse. The first vertex would remain fixed at the origin of the trackNode.

My thought is to transform the mouse coordinates in viewport space to the local coordinate screen space of the drawable to update the line endpoint as I drag the mouse around but I'm not sure if this is possible. Is there a transform that maps mouse coordinates into the local screen coordinates used by the trackNode for drawables?

I've also looked at using draggers to manipulate the line endpoint with the mouse but those only work with map coordinates.

Our application is using a Mercator projection mapNode.

Any help or suggestions on how I can solve this problem would be appreciated...

**Comment by gwaldron** [comment, marked as answer]
It sounds like you want to convert from the `TrackNode` coordinate space to viewport space. This is a standard MVPW (modelview->projection->window) transformation, which is constructed here:

https://github.com/gwaldron/osgearth/blob/master/src/osgEarth/ScreenSpaceLayoutDeclutter#L198

That matrix will transform from a "TrackNode local" coord to a viewport coord. The inverse will go in the other direction.

**Reply by keince** [reply]
Glenn,

Yes, that sounds like what I'm after.

Looking at the reference you provided, it looks like it is pulling the transforms from the camera found in osgUtils::Renderbin* bin by

```osg::Camera* cam = bin->getStage()->getCamera()```

Do I have access to this camera in the TrackNode or can I just get the camera from the view using

```osg::Camera* cam = view->getCamera();```

as usual?

**Comment by keince** [comment]
So...I tried Glenn's suggestion and it doesn't seem correct in transforming the TrackNode drawable vertex into Viewport CS and back.

Per the suggestion, here's the code I'm using to test. It attempts to transform the trackNode line from it's space into viewport space. For a sanity check it then transforms that result back into trackNode space...

```osg::Camera* cam = viewer->getCamera();
const osg::Matrix local2Viewport = cam->getViewMatrix() * cam->getProjectionMatrix() * cam->getViewport()->computeWindowMatrix();
osg::Matrix Viewport2local;
Viewport2local.invert(local2Viewport);
LineDrawable* leaderLine = static_cast<LineDrawable*>(_track->getDrawable("leader"));
if (leaderLine)
{
	osg::Vec3 leaderEnd = leaderLine->getVertex(1); // endpoint of line

	osg::Vec3d vplocation = leaderEnd * local2Viewport;
	osg::Vec3 tnlocation = Viewport2local * vplocation;
        std::cout << vplocation.x() << " " << vplocation.y() << std::endl;
	std::cout << tnlocation.x() << " " << tnlocation.y() << std::endl;
}
 ```
The line in the trackNode drawable is defined as (0,0 and (0, 50) for both endpoints. The window size is 800 x 600. The output is as follows transforming the point (0,50) into viewport space and back into trackNode space.
```
400 300.001
-0.00160123 -0.00120093
```
These results look wrong. I'd expect the transformed point into the viewport space to be somewhere other than the middle of the screen. I'd expect the transformed y-coordinate to be larger. The second line is just transforming back into TrackNode space which I expect it to be at (0,50).

Perhaps I'm not doing something correctly or something is missing?

**Comment by keince** [comment]
Glenn,

The suggestion didn't work. See my comment on the discussion topic.

Shayne

**Comment by keince** [comment]
So...just a follow up to my previous problem...

When I changed from a mapNode using a Mercator projection to a Geocentric mapNode, everything above works as expected with the world to window transformation.

I need to find out why the MVPW transformation doesn't work with Mercator...

Any suggestions or ideas on why using a mapNode with Mercator fails to give a proper world-to-window transform?

**Comment by keince** [comment]
A follow up to my previous follow up...

There were two problems.

1) I wasn't transforming the geopoint into the correct world coordinates, in this case Mercator.
2) I needed to set the perspective matrix to orthographic before applying the world-to-window transformation.

Now everything works.

Glenn's tip of using the MVPW transform helped me get the rest of the way through so thank you.

**Reply by nebsar** [reply]
Hi Keince,
Do you know how to change the heading of a tracknode or placenode symbol on the fly? Thanks.

**Reply by keince** [reply]
See my post on rotating the icon of a PlaceNode. Rotating the icon of a TrackNode is more tricky and more involved. I had to find the underlying geometry of the TrackNode icon image geometry and perform a rotation on that via an update callback.
